File: leetCode/array/top-k-frequent-elements.go
// https://leetcode.com/problems/top-k-frequent-elements/
type letterFrequency struct {
    letter int
    count int
}

func topKFrequent(nums []int, k int) []int {
    letterCount := map[int]int{}
    for _, letter := range nums {
        letterCount[letter]++
    }

    var lettersFrequency []letterFrequency
    for letter, count := range letterCount {
        lettersFrequency = append(lettersFrequency, letterFrequency{
            letter: letter,
            count: count,
        })
        for it := len(lettersFrequency) - 1; it > 0 && lettersFrequency[it].count > lettersFrequency[it - 1].count; it-- {
            lettersFrequency[it], lettersFrequency[it - 1] = lettersFrequency[it - 1], lettersFrequency[it]
        }
    }
    var resp []int
    for _, lc := range lettersFrequency {
        resp = append(resp, lc.letter)
        if len(resp) >= k {
            break
        }
    }
    return resp
}

File: leetCode/array/top-k-frequent-elements.py
# https://leetcode.com/problems/top-k-frequent-elements/
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        if len(nums) == 0:
            return []

        element_count = {}
        item = nums[0]
        count = 1
        for x in nums:
            element_count[x] = element_count.get(x, 0) + 1
            if element_count[x] > count:
                count = element_count[x]
                item = x
        
        resp = []
        for i in range(k):
            resp.append(item)
            del element_count[item]
            count = 0
            if len(element_count) == 0:
                break
            
            for x, x_count in element_count.items():
                if x_count > count:
                    count = x_count
                    item = x

        return resp
